### PR TITLE
feat(product): add uri resolver

### LINKED ISSUE #1471: add product uri resolver

### PR DESCRIPTION
**Pull request by griest024**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/graycoreio/daffodil/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] Other... Please describe:
```

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

## Other information
Fixes #1471

### CODE REVIEW

File: libs/product/state/src/resolvers/product-page/product-page.resolver.ts [hunk 1]
```diff
@@ -37,18 +16,10 @@ import { DaffProductReducersState } from '../../reducers/public_api';
 })
 export class DaffProductPageResolver implements Resolve<Observable<boolean>> {
```
Comment (lderrickable) [inline libs/product/state/src/resolvers/product-page/product-page.resolver.ts:17]: Should this be deprecated?

Comment (griest024) [reply libs/product/state/src/resolvers/product-page/product-page.resolver.ts:20]: yes, good catch

Comment (griest024) [reply libs/product/state/src/resolvers/product-page/product-page.resolver.ts:20]: changed

**Comment by lderrickable** [comment]
How exactly is this intended to be used? Because the app dev won't know what kind of url the user will give them (either domain/product/id or domain/product/product-name.html). I would expect that a common resolver would be needed to determine what kind of url is used, then that resolver could call productIdResolver.resolve or productUrlResolver.resolve. Is there another plan to handle this?

**Comment by griest024** [comment]
> a common resolver would be needed to determine what kind of url is used, then that resolver could call productIdResolver.resolve or productUrlResolver.resolve

That's exactly how it works: https://github.com/graycoreio/daffodil/blob/develop/libs/external-router/routing/src/guard/existence.guard.ts

**Comment by griest024** [comment]
Although the guard doesn't directly call the router, it adds a route to a specific place in the config and then that presumably is sufficient to invoke the correct resolver.

**Comment by griest024** [comment]
@lderrickable hang on I'll write a guide. I should have already done this TBH.

**Comment by griest024** [comment]
@lderrickable #1474

**Comment by lderrickable** [comment]
@griest024 With the setup outlined in the docs, how would it handle this route: `domain/product:id`? Seems like it would get to the `DaffProductPageUriResolver` and always result in a failed request.

Maybe something like this?
```
@NgModule({
  imports: [
    DaffExternalRouterModule.forRoot({
      failedResolutionPath: 'not-found',
    }, [
      {
        type: 'product:id',
        route: {
          component: MyProductPageComponent,
          ..., // additional route config, but without path
          resolve: {
            product: DaffProductPageIdResolver
          }
        }
      },
      {
        type: 'product/**',
        route: {
          component: MyProductPageComponent,
          ..., // additional route config, but without path
          resolve: {
            product: DaffProductPageUriResolver
          }
        }
      }
    ]),
  ],
})
class AppModule {}
```

**Comment by lderrickable** [comment]
actually, I'm not sure that would work, since the existence guard/resolver service would make the "resolvable route" a known route and therefore it wouldn't be caught by the `product/**`

**Comment by griest024** [comment]
@lderrickable I'm not sure exactly what your question is. If you're referring to the use case of putting routes in nested trees, that is supported by `insertionStrategy`. I added some better docs about that to my PR.

Note that `type` in the config is an enum defined by the platform and is returned in the driver reponse. For example Magento only ever has [these three](https://github.com/graycoreio/daffodil/blob/develop/libs/external-router/driver/magento/src/model/resolution-types.ts), You seem to be using it like its a route path but its not.

**Review by lderrickable — DISMISSED** [review]

**Comment by lderrickable** [comment]
lgtm

File: libs/product/state/src/resolvers/product-page-uri/product-page-uri.resolver.ts [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+})
+export class DaffProductPageUriResolver implements Resolve<Observable<boolean>> {
+  constructor(
+    @Inject(PLATFORM_ID) private platformId: string,
+    private store: Store<DaffProductReducersState>,
+    private dispatcher: ActionsSubject,
+  ) { }
+
+  resolve(route: ActivatedRouteSnapshot): Observable<boolean> {
+    this.store.dispatch(new DaffProductPageLoadByUrl(route.toString()));
```
Comment (damienwebdev) [inline libs/product/state/src/resolvers/product-page-uri/product-page-uri.resolver.ts:47]: Seeing that you mock `toString` in the test, do you know that `toString` returns what you expect? e.g. the path after the domain.

**Review by damienwebdev — CHANGES_REQUESTED** [review]

File: libs/product/state/src/resolvers/product-page-uri/product-page-uri.resolver.ts (hunk 2 above)
Comment (griest024) [reply libs/product/state/src/resolvers/product-page-uri/product-page-uri.resolver.ts]: fixed

**Review by damienwebdev — APPROVED** [review]

**Pull request merged**